### ISSUE TITLE
trainer.global_step print repetitive steps when there are more that one gradient_accumulation_steps 

### ISSUE THREAD

**Issue by shamanez** [PyTorchLightning/pytorch-lightning#6359]
I am using the latest pytorch_lighting version from conda.  Seems in this version, trainer.global_step change with gradient accumulation steps.  Also, the number of entire training steps. In the previous version, this wasn't the case. 

What is the correct approach if I want to do some task in every 5000 training steps? should I use self.trainer.global_step or batch_idx ?

When I use self.trainer.global_step actually repeats a number of times equal to the gradient accumulation steps.

**Comment by justusschock** [comment]
Hi,

You should use `self.trainer.global_step` since that one counts the backward passes, whereas `batch_idx` covers only the forward passes (there may be more forward passes then backward passes when accumulating gradients).

**Comment by shamanez** [comment]
Hi,

Thanks a lot. But I checked by printing self. trainer.global_step inside the training loop. The problem is then the counter value keeps repeating (equals to accumilate_gradieit_steps ). 

So then the function inside the If condition will also keep repeating.

**Comment by shamanez** [comment]
@justusschock 
something like this 

![image](https://user-images.githubusercontent.com/16892570/110149768-4d511480-7e43-11eb-889d-ee2e93e3cd05.png)

**Comment by justusschock** [comment]
@shamanez yes, but that's somewhat expected. Since you run multiple forwards for every backward. The number of forwards corresponds to the` batch_idx` while the number of backwards corresponds to `self. trainer.global_step`.

That means that inside your step, you'll have a different `batch_idx` everytime but `self. trainer.global_step` will only change after each backward, so you can have multiple forwards with the same value here. A whole step of the trainer is what we define as an arbitrary number of forwards and exactly one backward step.

**Comment by shamanez** [comment]
Ok  got it. I'm my case it would be better to use batch_index.
